pkg/distributor/push.go

```diff
@@ -36,24 +40,52 @@ type bufHolder struct {
 	buf []byte
 }
 
-var bufferPool = sync.Pool{
-	New: func() interface{} { return &bufHolder{buf: make([]byte, 256*1024)} },
-}
+var (
+	bufferPool = sync.Pool{
+		New: func() interface{} { return &bufHolder{buf: make([]byte, 256*1024)} },
+	}
+	errRetryBaseLessThanOneSecond    = errors.New("retry base duration should not be less than 1 second")
+	errNonPositiveMaxAllowedAttempts = errors.New("maxAllowedAttempts should be a positive value")
+)
 
 const (
 	SkipLabelNameValidationHeader = "X-Mimir-SkipLabelNameValidation"
 	statusClientClosedRequest     = 499
 )
 
+type RetryConfig struct {
+	Enabled            bool `yaml:"enabled" category:"experimental"`
+	Base               int  `yaml:"base_seconds" category:"experimental"`
+	MaxBackoffExponent int  `yaml:"max_backoff_exponent" category:"experimental"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet.
+func (cfg *RetryConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, "distributor.retry-after-header.enabled", false, "Enabled controls inclusion of the Retry-After header in the response: true includes it for client retry guidance, false omits it.")
+	f.IntVar(&cfg.Base, "distributor.retry-after-header.base-seconds", 3, "Base duration in seconds for calculating the Retry-After header in responses to 429/5xx errors.")
```

Comment:
Do we want to default the base Retry-After to 3 seconds? That seems a bit long for an initial retry attempt.

Reply:
We could use 1 as a "base" instead, then min retry would be 1 second which is also minimum we can specify in Retry-After header.

I find it bit confusing to call  this parameter "base", since we use "2" as exponent base. We could rename this to minimum retry period, "factor", or just drop this parameter completely (I'd be in favor of that).

Reply:
Re: naming, it's common to call this something like "initial delay" or "min backoff", since subsequent delays or computed backoffs are larger.

Reply:
It is for a better randomness at edge cases. Since agent would already retry many times when internet reestablished. they will go directly all 30s. Although we've extensively debated this, in practice, a majority of users may opt to retain the default configuration. I propose affording this configuration an opportunity for experimentation and evaluation.

Reply:
The idea with a max delay is that you could still introduce jitter on top of it, as your current implementation does. But jitter could be introduced and even configured, if desired, separately (ex: jitter factor = .1).